acme/dns_challenge.go

```diff
@@ -62,48 +64,64 @@ func (s *dnsChallenge) Solve(chlng challenge, domain string) error {
 
 	fqdn, _, _ := DNS01Record(domain, keyAuth)
 
-	preCheckDNS(domain, fqdn)
+	if err := preCheckDNS(fqdn); err != nil {
+		log.Printf("Error checking DNS record: %v", err)
+	}
 
 	return s.validate(s.jws, domain, chlng.URI, challenge{Resource: "challenge", Type: chlng.Type, Token: chlng.Token, KeyAuthorization: keyAuth})
 }
 
-func checkDNS(domain, fqdn string) bool {
-	// check if the expected DNS entry was created. If not wait for some time and try again.
-	m := new(dns.Msg)
-	m.SetQuestion(domain+".", dns.TypeSOA)
-	c := new(dns.Client)
-	in, _, err := c.Exchange(m, "google-public-dns-a.google.com:53")
+// Check if the expected DNS record was created. If not wait for some time and try again.
+// We do a semi-recursive query starting at the top-level domain.
+func checkDNSrecord(fqdn string) error {
+	labels := dns.SplitDomainName(fqdn)
+	gTld := labels[len(labels)-1]
+	nss, err := net.LookupNS(gTld)
 	if err != nil {
-		return false
+		return err
 	}
+	authoritativeNs := nss[0].Host
 
-	var authorativeNS string
-	for _, answ := range in.Answer {
-		soa := answ.(*dns.SOA)
-		authorativeNS = soa.Ns
-	}
+	c := new(dns.Client)
 
 	fallbackCnt := 0
+	recursionsCnt := 0
 	for fallbackCnt < preCheckDNSFallbackCount {
+	Next:
+		m := new(dns.Msg)
 		m.SetQuestion(fqdn, dns.TypeTXT)
-		in, _, err = c.Exchange(m, authorativeNS+":53")
+		m.RecursionDesired = false
+
+		recursionsCnt++
+		in, _, err := c.Exchange(m, authoritativeNs+":53")
 		if err != nil {
-			return false
+			return err
 		}
 
 		if len(in.Answer) > 0 {
-			return true
+			for _, rr := range in.Answer {
+				if _, ok := rr.(*dns.TXT); ok {
+					return nil
+				}
+			}
 		}
 
-		fallbackCnt++
-		if fallbackCnt >= preCheckDNSFallbackCount {
-			return false
+		if len(in.Ns) > 0 {
+			for _, r := range in.Ns {
+				rr, ok := r.(*dns.NS)
+				if ok && recursionsCnt < preCheckDNSMaxRecursions {
```

Comment:
Guess we should query all NS at some point, right? Instead of artificially limit the number to five.